docs: add form customization docs (#117)
aerogear/OpenVolunteerPlatform: docs/forms.md

---
id: adding-and-customizing-forms
title: Out of the box Form
---

# 1. Introduction

> Open Volunteer uses [`uniforms`](https://www.npmjs.com/package/uniforms) and [`unifomrs-ionic`](https://www.npmjs.com/package/uniforms-ionic) packages to create html forms with built-in state management and validatio with minimum code. 

Reviewer: ```suggestion
Open Volunteer uses [`uniforms`](https://www.npmjs.com/package/uniforms) and [`uniforms-ionic`](https://www.npmjs.com/package/uniforms-ionic) packages to create html forms with built-in state management and validation with minimum code. 
```

> You usually woould not have to install any of these libraries since the reference implementation comes with them already. 

Reviewer: ```suggestion
You usually would not have to install any of these libraries since the reference implementation comes with them already. 
```

> Thanks to this integration, Open Volunteer Platform you with simple re-usable form components which allows for rapid prototyping and cleaner React components. This source code for the forms is located in `client/src/forms` and `client-admin/src/forms`

Reviewer: ```suggestion
Thanks to this integration, Open Volunteer Platform provides you with simple re-usable form components which allows for rapid prototyping and cleaner React components. This source code for the forms is located in `client/src/forms` and `client-admin/src/forms`
```

# 2. Modifying an existing form

> In this section we'll see how we'll are going to add additional information, add validation, or remove informations from an existing form.

Reviewer: ```suggestion
In this section we'll see how to add additional information, add validation, or remove information from an existing form.
```

### Removing a field 

This can be achieved by deleting the field from the form definition object. 

### Adding validation to postcode field

Let's asssume that we want our volunteer's postcodes can be between `1000-9999`, we can add this sort of validation using the min/max boundaries

```ts
{
    postcode: {
        type: Number,
        required: false,
        max: 9999,
        min: 1000
    }
}
```

## 3. Adding a new form

The process of adding a new custom form requires two steps. 

1. Defining the schema
2. Using the schema in a React Component 

Let's take a look at an example by trying to add a form to display product full details in the `client` application.

### Start by defining the schema. 

In the `client/src/forms/product.ts` file, let's define the following schema.

```ts
import SimpleSchema from '../config/SimpleSchema';
import SimpleSchema2Bridge from 'uniforms-bridge-simple-schema-2';

const productSchema = new SimpleSchema({
    label: {
        type: String,
        max: 50,
        readonly: true
    },
    description: {
        type: String,
        max: 500,
        readonly: true
    }
} as any);

return new SimpleSchema2Bridge(productSchema);

```

> The schema above will enable us to display full product informations which cannot be edited thanks to the `readonly` property. 

Reviewer: ```suggestion
The schema above will enable us to display full product information which cannot be edited thanks to the `readonly` property. 
```
